lua/oatmeal.lua

```diff
@@ -1,56 +1,32 @@
+local utils = require("utils")
+
 local M = {}
 
 local config = {}
 local session_context = {}
 
-local function to_base64(data)
-  local b = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
-  return (
-    (data:gsub(".", function(x)
-      local r, b = "", x:byte()
-      for i = 8, 1, -1 do
-        r = r .. (b % 2 ^ i - b % 2 ^ (i - 1) > 0 and "1" or "0")
-      end
-      return r
-    end) .. "0000"):gsub("%d%d%d?%d?%d?%d?", function(x)
-      if #x < 6 then
-        return ""
-      end
-      local c = 0
-      for i = 1, 6 do
-        c = c + (x:sub(i, i) == "1" and 2 ^ (6 - i) or 0)
-      end
-      return b:sub(c + 1, c + 1)
-    end) .. ({ "", "==", "=" })[#data % 3 + 1]
-  )
-end
-
 function M.set_context()
   local start_line = vim.api.nvim_win_get_cursor(0)[1]
   local end_line = nil
   local code = ""
 
-  -- Visual Mode
-  local vstart = vim.fn.getpos("'<")
-  local vend = vim.fn.getpos("'>")
-  if vstart ~= nil and vend ~= nil and vend[2] ~= 0 then
-    start_line = vstart[2]
-    end_line = vend[2]
+  if utils.mode_is_visual() then
+    local selection
+    code, selection = utils.get_visual_selection()
 
-    local select_start_line = start_line
-    if select_start_line ~= 0 then
-      select_start_line = select_start_line - 1
+    if not selection then
+      return
     end
 
-    local lines = vim.api.nvim_buf_get_lines(0, select_start_line, end_line, true)
-    code = table.concat(lines, "\n")
+    start_line = selection.start.line
+    end_line = selection["end"].line
   end
 
   session_context = {
     file_path = vim.fn.expand("%:p"),
     language = vim.bo.filetype,
     buffer_id = vim.fn.bufnr(),
-    code = to_base64(code),
+    code = vim.base64.encode(code),
```

Comment:
Love this, thanks! I'd gladly accept this in another PR while figuring out selection.